File: src/main/java/cn/hirrao/javaee/Controller/UserController.java
package cn.hirrao.javaee.Controller;


import cn.hirrao.javaee.Entity.Result;
import cn.hirrao.javaee.Entity.User;
import cn.hirrao.javaee.Service.UserService;
import cn.hirrao.javaee.Utils.JwtUtil;
import cn.hirrao.javaee.Utils.SnowFlake;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.util.DigestUtils;
import org.springframework.web.bind.annotation.Mapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/user")
public class UserController {

    @Autowired(required = false)
    private UserService userService;

    SnowFlake snowFlake = new SnowFlake(1, 1);
    @PostMapping("/register")
    public Result register(String userName,String userPassword) {
        if(userName == "" || userPassword == "") {
            return Result.error("用户名或密码不能为空");
        } else if(userName.length() < 3 || userName.length() > 20) {
            return Result.error("用户名长度必须在3-20之间");
        } else if(userPassword.length() < 6 || userPassword.length() > 20) {
            return Result.error("密码长度必须在6-20之间");
        } else if (!userName.matches("^[a-zA-Z0-9_]{3,20}$")||!userPassword.matches("^[a-zA-Z0-9_]{6,20}$")) {
            return Result.error("用户名或密码只能包含字母、数字和下划线");
        }
        User u = userService.findByUsername(userName);
        if (u == null) {
            long uid = snowFlake.nextId();
            userService.register(uid,userName, userPassword);
            return Result.success();
        } else {
            return Result.error("用户名已被占用");
        }
    }

    @PostMapping("/login")
    public Result login(String userName, String userPassword) {
        if(userName == "" || userPassword == "") {
            return Result.error("用户名或密码不能为空");
        }
        User u = userService.findByUsername(userName);
        if(u == null) {
            return Result.error("用户名不存在");
        } else {
            String p1 = DigestUtils.md5DigestAsHex(userPassword.getBytes());
            String p2 = DigestUtils.md5DigestAsHex(p1.getBytes());
            if(p2.equals(u.getUserPassword())) {
                //密码正确，根据用户的uid和用户名生成token
                Map<String,Object> claims = new HashMap<>();
                claims.put("uid", u.getId());
                claims.put("userName", u.getUserName());
                String token = JwtUtil.genToken(claims);
                return Result.success(token);
            } else {
                return Result.error("密码错误");
            }
        }
    }
}

File: api/src/main/kotlin/cn/hirrao/javaee/controller/UserController.kt
package cn.hirrao.javaee.controller;

import cn.hirrao.javaee.entity.Result;
import cn.hirrao.javaee.service.UserService;
import cn.hirrao.javaee.utils.ThreadLocalUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/user")
public class UserController {

    private final Logger logger = LoggerFactory.getLogger(getClass());
    private final UserService userService;

    @Autowired
    public UserController(UserService userService) {
        this.userService = userService;
    }

    @GetMapping("/userInfo")
    public Result userInfo() {
        var user = ThreadLocalUtil.get();
        logger.debug("获取用户信息{}", user);
        return Result.success(user);
    }

    @PutMapping("/update")
    public Result update(@RequestBody Map<String, String> map) {
        logger.debug("更新用户信息{}", map);
        var user = ThreadLocalUtil.get();
        var uid = user.getUid();
        var phoneNumber = map.get("phoneNumber");
        var sex = map.get("sex");
        var birthday = map.get("birthday");
        userService.update(uid, phoneNumber, sex, birthday);
        return Result.success();
    }

    @PostMapping("/accountsInfo")
    public Result accountsInfo(@RequestBody Map<String, String> map) {
        var curPage = Integer.parseInt(map.get("curPage"));
        var size = Integer.parseInt(map.get("size"));
        return Result.success(userService.accountsInfo(curPage, size));
    }

    @PostMapping("/modifyUserInfo")
    public Result modifyUserInfo(@RequestBody Map<String, String> map) {
//        User user = ThreadLocalUtil.get();
//        var uid = user.getUid();
        var uid = Long.parseLong(map.get("uid"));
        var userName = map.get("userName");
        var phoneNumber = map.get("phoneNumber");
        var sex = map.get("sex");
        var birthday = map.get("birthday");
        var permission = Integer.parseInt(map.get("permission"));
        System.out.println("modifyUserInfo uid:" + uid);
        System.out.println("modifyUserInfo userName:" + userName);
        System.out.println("modifyUserInfo sex:" + sex);
        System.out.println("modifyUserInfo birthday:" + birthday);
        System.out.println("modifyUserInfo permission:" + permission);

        userService.modifyUserInfo(uid, userName, phoneNumber, sex, birthday, permission);
        return Result.success();
    }

    @PostMapping("/updateUserName")
    public Result updateUserName(@RequestBody Map<String, String> map) {
        var user = ThreadLocalUtil.get();
        var uid = user.getUid();
        var userName = map.get("userName");
        userService.updateUserName(uid, userName);
        return Result.success();
    }

    @PostMapping("/updateSex")
    public Result updateSex(@RequestBody Map<String, String> map) {
        var user = ThreadLocalUtil.get();
        var uid = user.getUid();
        var sex = map.get("sex");
        userService.updateSex(uid, sex);
        return Result.success();
    }

    @PostMapping("/updateBirthday")
    public Result updateBirthday(@RequestBody Map<String, String> map) {
        var user = ThreadLocalUtil.get();
        var uid = user.getUid();
        var birthday = map.get("birthday");
        userService.updateBirthday(uid, birthday);
        return Result.success();
    }

    @PostMapping("/deleteUser")
    public Result deleteUser(@RequestBody Map<String, String> map) {
//        System.out.println(map.size());
//        User user = ThreadLocalUtil.get();
        var uid = Long.parseLong(map.get("uid"));
        userService.deleteUser(uid);
        System.out.println("delete uid:" + uid);
        return Result.success();
    }

    @PostMapping("/searchUserByCondition")
    public Result searchUserByCondition(@RequestBody Map<String, String> map) {
        var curPage = Integer.parseInt(map.get("curPage"));
        var size = Integer.parseInt(map.get("size"));
        var searchCondition = map.get("searchCondition");
        var conditionValue = map.get("conditionValue");
        System.out.println("curPage:" + curPage);
        System.out.println("size:" + size);
        System.out.println("searchCondition:" + searchCondition);
        System.out.println("conditionValue:" + conditionValue);
        if (!searchCondition.isEmpty() && !conditionValue.isEmpty()) {
            return Result.success(userService.searchUserByCondition(curPage, size, searchCondition, conditionValue));
        } else {
            return Result.success(userService.accountsInfo(curPage, size));
        }
    }
}
